server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -407,22 +406,56 @@ public long compareAndExchangeRegister(String key, long expected, long updated)
             if (found == expected) {
                 buf.clear().putLong(updated).flip();
                 while (buf.remaining() > 0) {
-                    channel.write(buf, buf.position());
+                    fileChannel.write(buf, buf.position());
                 }
-                channel.force(true);
+                fileChannel.force(true);
             }
             return found;
+        } catch (OverlappingFileLockException e) {
+            throw new ConcurrentRegisterOperationException(e);
         }
     }
 
-    private static FileChannel openOrCreateAtomic(Path path) throws IOException {
-        try {
-            if (Files.exists(path) == false) {
-                return FileChannel.open(path, StandardOpenOption.READ, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+    private record LockedFileChannel(FileChannel fileChannel, Closeable fileLock) implements Closeable {
+
+        // avoid concurrently opening/closing locked files
```

Comment:
Can we add to the comment that we're serializing the `LockedFileChannel` creation due to a tripping assertion in the JDK?

Reply:
Sure, done in 37e37615c46.